src/solvers/dgsem_tree/indicators_2d.jl

```diff
@@ -223,5 +223,476 @@ function (indicator_max::IndicatorMax)(u::AbstractArray{<:Any,4},
+end
+
+# this method is used when the indicator is constructed as for AMR
+function create_cache(typ::Type{IndicatorNNPP}, mesh, equations::AbstractEquations{2}, dg::DGSEM, cache)
+  create_cache(typ, equations, dg.basis)
+end
+
+
+function (indicator_nnpp::IndicatorNNPP)(u::AbstractArray{<:Any,4},
+                                                   equations, dg::DGSEM, cache;
+                                                   kwargs...)
+  @unpack alpha_max, alpha_min, alpha_smooth, alpha_continuous, alpha_amr, variable, network = indicator_nnpp
+  @unpack alpha, alpha_tmp, indicator_threaded, modal_threaded, modal_tmp1_threaded = indicator_nnpp.cache
+  # TODO: Taal refactor, when to `resize!` stuff changed possibly by AMR?
+  #       Shall we implement `resize!(semi::AbstractSemidiscretization, new_size)`
+  #       or just `resize!` whenever we call the relevant methods as we do now?
+  resize!(alpha, nelements(dg, cache))
+  if alpha_smooth
+    resize!(alpha_tmp, nelements(dg, cache))
+  end
+
+  @threaded for element in eachelement(dg, cache)
+    indicator  = indicator_threaded[Threads.threadid()]
+    modal      = modal_threaded[Threads.threadid()]
+    modal_tmp1 = modal_tmp1_threaded[Threads.threadid()]
+
+    # Calculate indicator variables at Gauss-Lobatto nodes
+    for j in eachnode(dg), i in eachnode(dg)
+      u_local = get_node_vars(u, equations, dg, i, j, element)
+      indicator[i, j] = indicator_nnpp.variable(u_local, equations)
+    end
+
+    # Convert to modal representation
+    multiply_scalar_dimensionwise!(modal, dg.basis.inverse_vandermonde_legendre, indicator, modal_tmp1)
+
+    # Calculate total energies for all modes, without highest, without two highest
+    total_energy = zero(eltype(modal))
+    for j in 1:nnodes(dg), i in 1:nnodes(dg)
+      total_energy += modal[i, j]^2
+    end
+    total_energy_clip1 = zero(eltype(modal))
+    for j in 1:(nnodes(dg)-1), i in 1:(nnodes(dg)-1)
+      total_energy_clip1 += modal[i, j]^2
+    end
+    total_energy_clip2 = zero(eltype(modal))
+    for j in 1:(nnodes(dg)-2), i in 1:(nnodes(dg)-2)
+      total_energy_clip2 += modal[i, j]^2
+    end
+    total_energy_clip3 = zero(eltype(modal))
+    for j in 1:(nnodes(dg)-3), i in 1:(nnodes(dg)-3)
+      total_energy_clip3 += modal[i, j]^2
+    end
+
+    
+    # Calculate energy in lower modes and polynomial degree for the network input
+    X = zeros(4,1)
+    X[1] = (total_energy - total_energy_clip1)/total_energy
+    X[2] = (total_energy_clip1 - total_energy_clip2)/total_energy_clip1
+    X[3] = (total_energy_clip2 - total_energy_clip3)/total_energy_clip2 
+    X[4] = nnodes(dg) 
+
+    # Scale input data
+    X = X ./max(maximum(abs.(X)),1)
+
+    if alpha_continuous && !alpha_amr
+      # Set good cells to 0 and troubled cells to continuous value of the network prediction
+      if network(X)[1] > 0.5
+        alpha_element = network(X)[1]
```

Comment:
This will evaluate the network twice, correct? Maybe better to use something like
```suggestion
      probability_troubled_cell = network(X)[1] 
      if probability_troubled_cell > 0.5
        alpha_element = probability_troubled_cell
```